lib/elixir/test/elixir/exception_test.exs

```diff
@@ -477,16 +477,37 @@ defmodule ExceptionTest do
              """
     end
 
-    test "annotates badarg on apply" do
+    test "annotates dot syntax on invalid argument errors" do
       assert blame_message([], & &1.foo) ==
-               "you attempted to apply a function named :foo on []. If you are using Kernel.apply/3, make sure " <>
-                 "the module is an atom. If you are using the dot syntax, such as " <>
-                 "map.field or module.function(), make sure the left side of the dot is an atom or a map"
+               "could not find key :foo on []. When using the map.field syntax, " <>
+                 "make sure the left side of the dot is a map"
+
+      assert blame_message([], & &1.foo()) ==
+               "could not find function :foo on []. When using the module.function() " <>
+                 "syntax, make sure the left side of the dot is a module"
+
+      # atoms that are invalid module name:
+      assert blame_message(nil, & &1.foo()) ==
+               "could not find function :foo on nil. When using the module.function() " <>
+                 "syntax, make sure the left side of the dot is a module"
+
+      assert blame_message(false, & &1.foo()) ==
+               "could not find function :foo on false. When using the module.function() " <>
+                 "syntax, make sure the left side of the dot is a module"
+
+      assert blame_message(nil, fn _ -> nil.foo() end) ==
+               "could not find function :foo on nil. When using the module.function() " <>
+                 "syntax, make sure the left side of the dot is a module"
```

Comment:
- this is actually going in a different branch as `blame_message(nil, & &1.foo())`, which is why I added a test
- this triggers a compile-time warning, any idea how I could silence it? `warning: nil.foo/0 is undefined`

Reply:
You need to use eval.